dashboard/app/controllers/script_levels_controller.rb

```diff
@@ -247,15 +247,35 @@ def stage_extras
       return
     end
 
-    @stage = Script.get_from_cache(params[:script_id]).lesson_by_relative_position(params[:stage_position].to_i)
+    user = @user || current_user
+    @stage = Script.get_from_cache(
+      params[:script_id]
+    ).lesson_by_relative_position(
+      params[:stage_position].to_i
+      )
     @script = @stage.script
+    script_bonus_levels_by_stage = @script.get_bonus_script_levels(@stage)
+
+    # bonus level summaries explicitly don't contain any user-specific data,
+    # so we need to merge in the user's progress.
+    script_bonus_levels_by_stage.each do |stage|
+      stage[:levels].each do |level_summary|
+        ul = UserLevel.find_by(
```

Comment:
We sometime return more than 10+ bonus levels at once, an example: https://studio.code.org/s/coursef-2020/stage/19/extras (there are arrows to navigate among bonus levels). Could it cause a performance hit by making one DB query per level here?

Reply:
this query was already happening once per level in `script_level.rb` line 497, and is necessary in order to show users' progress on the levels. this will cause a slight performance hit since previously the results of those queries were cached, but that's specifically the bug that this PR is fixing -- since the results were cached, we were showing the wrong progress data (whatever data happened to be cached, rather than the current user's own data).

so while this change does result in a performance hit, it's a necessary one, and in my testing i have not seen a noticeable degradation in page load time.

Reply:
reducing DB queries is also important for maintaining database uptime during peak traffic, though I'm not sure if this is still a big concern now that we've moved to aurora. in general our strategy is to cache HOC scripts (which do not use lesson extras anyway) so that these types of requests do not hit the server or DB on each request. all this to say it seems ok to me to add a few DB queries here.

Reply:
Thank you both for the explanation; I'm okay with that. 
In case it causes performance hit later, combining all those individual queries into 1 may help `UserLevel.find_by(user_id: user.id, script: @script, level_id: <an_array_of_level_ids>)`.